File: solutions/0278. First Bad Version/0278.cs
// Time complexity: O(logn)
// Space complexity: O(1)
public class Solution : VersionControl {
    public int FirstBadVersion(int n) {
        int good = 0;
        int bad = n;

        while (bad > good + 1) {
            int mid = good + ((bad - good) / 2);
            if (IsBadVersion(mid)) {
                bad = mid;
            } else {
                good = mid;
            }
        }
        return bad;
    }
}

File: solutions/0278. First Bad Version/0278.py
# Time complexity: O(logn)
# Space complexity: O(1)
class Solution:
    def firstBadVersion(self, n: int) -> int:
        good = 0
        bad = n

        while bad > good + 1:
            mid = good + ((bad - good) // 2)
            if isBadVersion(mid):
                bad = mid
            else:
                good = mid

        return bad 